File: lab7-8/SMS_5_6_2/SMS_7_8/dmc_dysk/DMC.c
const short int Nu= 4; //horyzont sterowania
const short int N = 4; //horyzont predykcji
const short int lambda = 0;
const short int D =2; //horyzont dynamiki

double deltaU =0;
double deltaUP[D-1];
double s[30];
double Ku[D-1]; //tu wyniki z matlaba
double Ke;
float DMC(float u,float y)
{	
	double e = 0, Ku_deltaUP =0;
	e = y_zad - y;
	for(int i = 0; i< D-1 ;i++)
		Ku_deltaUP += Ku[i]*deltaUP[i];
	deltaU = Ke*e - Ku_deltaUP;
	
	for(int i=D; i>0 ;i--)
		deltaUP[i]= deltaUP[i-1];
	deltaUP[0] = deltaU;
	
	return u_prev1 +deltaU;
}

File: lab7-8/SMS_5_6_2/SMS_7_8/dmc_dysk/DMC.m
lambda=1; %parametr lambda np. 1
D=12; %horyzont dynamiki (D)
N=12;%horyzont predykcji (N)
Nu=6; %horyzont sterowania (Nu)(ilosc przyszlych przyrostow wartosci sterowania)

s=[1 2 3 4 5 6 7 8 9 10 11 12];
%macierz wspó³czynników odpowiedzi skokowej wymiary(NxNu)
M=zeros(N,Nu); 
for i=1:N
 for j=1:Nu
  if (j<=i)             %wypelnianie macierzy trojkatnej dolnej M   
   M(i,j)=s(i-j+1);
  end;
 end;
end;

I=eye(Nu);              %tworzenie macierzy jednostkowej o wymiarach NuxNu
K=(M'*M+lambda*I)\M';   %macierz K
Mp=zeros(N,D-1);        %macierz ma wymiary Nx(D-1)
%wypelnianie macierzy Mp
for i=1:N
 for j=1:D-1
  if i+j<=D
   Mp(i,j)=s(i+j)-s(j);
  else
   Mp(i,j)=s(D)-s(j);
  end;      
 end;
end;
Ke=sum(K(1,:));         %wspó³czynnik Ke
Ku=K(1,:)*Mp;           %wspó³czynnik Ku
